lib/grape/dsl/parameters.rb

```diff
@@ -199,6 +196,42 @@ def params(params)
         end
         params
       end
+
+      # @param params [Hash] initial hash of parameters
+      # @return hash of parameters with index information relevant for the current scope
+      # example:
+      # params: [{"name"=>"Job", "parents"=>[{"name"=>"Joy"}, {"name"=>"Bob"}], "_param_index": {"children_index": 0}},
+      #   {"name"=>"Jim", "parents"=>[{}], "_param_index": {"children_index": 1}}]
+      # element: parents
+      # result: [{"name"=>"Joy", "_param_index": {children[parents]: 0, "children": 0}},
+      #   {"name"=>"Bob", "_param_index": {children[parents]: 1, "children": 0}},
+      #   { "_param_index": {children[parents]: 0, "children": 1}}]
+      def params_with_index(params)
+        params = @parent.params_with_index(params) if @parent
+        if @element
+          if params.is_a?(Array)
+            params = params.map.with_index do |el, index|
+              el[@element] ? convert_to_param_index(el[@element], el['_param_index'], index) : {}
+            end.flatten
+          elsif params.is_a?(Hash)
+            params = params[@element] ? convert_to_param_index(params[@element], params['_param_index']) : {}
+          else
+            params = {}
```

Comment:
params method return hash of parameters relevant for the current scope. Because index information of array will not be retained, when the validation fails, error information is not complete. params_with_index will retain index information of array, when the validation fails to get the complete message by full_message method.

example：	
I defined the following parameters：
```ruby
subject.params do
    group :children, type: Array do
       requires :name
       group :parents, type: Array do
          requires :name
       end
    end
end
```
if param is:
```ruby
children: [
          { name: 'Job', parents: [{ name: 'Joy' }, {name: 'Bob'}] },
          { name: 'Jim', parents: [{}] }
]
```
expect raise 'children[1][parents][0][name] is missing'，got: "children[0][parents][0][name] is missing"